libs/wingc/src/struct_schema.rs

```diff
@@ -0,0 +1,148 @@
+			}
+			Type::Struct(ref s) => {
+				let mut code = CodeMaker::default();
+				code.open("{");
+				code.line("type: \"object\",");
+				code.open("properties: {");
+				code.add_code(self.get_struct_env_properties(&s.env));
+				code.close("},");
+				code.add_code(self.get_struct_schema_required_fields(&s.env));
+				code.close("}");
+				code.to_string().strip_suffix("\n").unwrap().to_string()
+			}
+			Type::Array(ref t) | Type::Set(ref t) => {
+				let mut code = CodeMaker::default();
+				code.open("{");
+
+				code.line("type: \"array\",");
+
+				if matches!(**typ, Type::Set(_)) {
+					code.line("uniqueItems: true,");
+				}
+
+				code.line(format!("items: {}", self.get_struct_schema_field(t, env)));
+
+				code.close("}");
+				code.to_string().strip_suffix("\n").unwrap().to_string()
+			}
+			Type::Map(ref t) => {
+				let mut code = CodeMaker::default();
+				code.open("{");
+
+				code.line("type: \"object\",");
+				code.line(format!(
+					"patternProperties: {{ \".*\": {} }}",
+					self.get_struct_schema_field(t, env)
+				));
+
+				code.close("}");
+				code.to_string().strip_suffix("\n").unwrap().to_string()
+			}
+			Type::Optional(t) => self.get_struct_schema_field(&t, env),
+			Type::Json(_) => "{ type: \"object\" }".to_string(),
+			_ => "{ type: \"null\" }".to_string(),
+		}
+	}
+
+	fn get_struct_env_properties(&self, env: &SymbolEnv) -> CodeMaker {
+		let mut code = CodeMaker::default();
+		for (field_name, (.., kind)) in env.symbol_map.iter() {
+			code.line(format!(
+				"{}: {},",
+				field_name,
+				self.get_struct_schema_field(&kind.as_variable().unwrap().type_, env)
+			));
+		}
+		code
+	}
+
+	fn get_struct_schema(&self, struct_: &Struct) -> CodeMaker {
+		let mut code = CodeMaker::default();
+
+		code.open("{");
+		code.line(format!("id: \"/{}\",", struct_.name));
+		code.line("type: \"object\",".to_string());
+
+		code.open("properties: {");
+
+		code.add_code(self.get_struct_env_properties(&struct_.env));
+
+		//close properties
+		code.close("},");
+
+		code.add_code(self.get_struct_schema_required_fields(&struct_.env));
+
+		// close schema
+		code.close("}");
+		code
+	}
+}
+
+// Looks for any references to a struct type, and emits a struct schema file for it.
+impl<'a> Visit<'a> for StructSchemaVisitor<'a> {
+	fn visit_reference(&mut self, node: &'a crate::ast::Reference) {
+		match node {
+			Reference::InstanceMember { .. } => {
+				visit::visit_reference(self, node);
+			}
+			Reference::TypeMember { type_name, .. } => {
+				if is_udt_struct_type(type_name, self.ctx.current_env().unwrap()) {
+					let type_ = resolve_user_defined_type(type_name, self.ctx.current_env().unwrap(), 0);
+					let struct_code = self.get_struct_schema(type_.unwrap().as_struct().unwrap());
+
+					// add the schema to the jsifier's referenced struct schemas
+					self
+						.jsify
+						.add_referenced_struct_schema(type_name.clone().to_string(), struct_code);
```

Comment:
I'm down with this. But would like to say that what we're doing here is assigning a schema to a type. This, again, isn't strictly related to the `JSifier`. We could have just as well, I think, added this map to `Types`. Imagine our type-system having a schema per struct regardless of the `JSifier`.